### DISCUSSION TITLE
Mutation does not work in MixedVariableGA

### DISCUSSION THREAD

**Discussion by mhr-azizi** [anyoptimization/pymoo#624] [Q&A]
I have written a code for multi objective optimization using pymoo. I have two cost function of six variables which the first variable is integer, and the others are float. So, I used MixedVariableGA for minimization. I used seed=1 to have a constant initial population and evaluate the effects of mutation on the results. However, the results are constant in every run, no matter how the mutation parameters changes. This is a part of code I developed:

`

    class Multi_Objective_Problem(ElementwiseProblem):
        
        
        def __init__(self, cost_fun_evaluator, **kwargs):
            
            self.cost_fun_evaluator = cost_fun_evaluator
            x1_L = 10       # Integer variable
            x1_U = 500      # Integer variable
            x2_L = 1.0
            x2_U = 8.0 
            x3_L = 2.0
            x3_U = 9.0                 
            x4_L = 2.0
            x4_U = 9.0
            x5_L = 2.0
            x5_U = 30.0
            x6_L = 2.0
            x6_U = 30.0
            vars = {'x1': Integer(bounds=(x1_L, x1_U)),
                    'x2': Real(bounds=(x2_L, x2_U)),
                    'x3': Real(bounds=(x3_L, x3_U)),
                    'x4': Real(bounds=(x4_L, x4_U)),
                    'x5': Real(bounds=(x5_L, x5_U)),
                    'x6': Real(bounds=(x6_L, x6_U))}
            
            super().__init__(n_obj=2, n_ieq_constr=0, vars=vars)
                             
                             
            
            
        def _evaluate(self, X, out, *args, **kwargs):
            x1, x2, x3, x4, x5, x6 = X["x1"], X["x2"], X["x3"], X["x4"], X["x5"], X["x6"]
            solution = [x1, x2, x3, x4, x5, x6]
            y1, y2 = self.cost_fun_evaluator.objective_function(solution)
            out["F"] = [y1, y2]
            
            
            
     
    problem = Multi_Objective_Problem(cost_fun_evaluator=cost_fun)
    mutation = PolynomialMutation(prob=0.25, eta=8)
    selection = RandomSelection()
    crossover = TwoPointCrossover(prob=0.9)
    
    algorithm = MixedVariableGA(pop_size=110,
                                survival= RankAndCrowding(),
                                mutation=mutation,
                                selection=selection,
                                crossover=crossover)
    
    # Run GA:
    result = minimize(problem, 
                      algorithm, ('n_gen', 100),
                      seed=1,
                      verbose=True)
`

The results are constant when I change prob and eta for PolynomialMutation. In addition, using eliminate_duplicates=True leads to following error in my code:

TypeError: float() argument must be a string or a real number, not 'dict'

**Comment by blankjul** [comment]
This fails by design because `PolynomialMutation` only supports float variables (as proposed in the paper). For mixed variables I recommend sticking to the tutorial here: 
https://pymoo.org/customization/mixed.html

**Reply by mhr-azizi** [reply]
Thank you for your response and help.
I read the tutorial you mentioned. However, I didn't see any description about how mutation should be handled in mixed variables problem, when some variables are integers. Do you mean I have to write a custom mutation class, as describe in "Custom Variable Type" tutorial at https://pymoo.org/customization/custom.html ?
Is there other type of mutation defined in pymoo library (such as Bitflip mutation) that can be used in my problem? 
Many thanks,

**Comment by blankjul** [comment]
Checkout `https://github.com/anyoptimization/pymoo/blob/main/pymoo/core/mixed.py` where the operators are defined:

```

        if crossover is None:
            crossover = {
                Binary: UX(),
                Real: SBX(),
                Integer: SBX(vtype=float, repair=RoundingRepair()),
                Choice: UX(),
            }

        if mutation is None:
            mutation = {
                Binary: BFM(),
                Real: PM(),
                Integer: PM(vtype=float, repair=RoundingRepair()),
                Choice: ChoiceRandomMutation(),
            }

```

Depending on the type the corresponding default operator is used. Feel free to provide your own operator ensemble to further customize.

**Reply by mhr-azizi** [reply]
Thank you for the reply. 
I used the code sample you provided to define the MixedVariableMating class and used it in defining the algorithm. However, the algorithm slows down significantly for my optimization problem, making it unapplicable. In addition, I also defined the mutation operator as the above dictionary in my code to use the mutation based on the type of variables (following code). However, this has not any effect on the results of the algorithm. 
 Is there any more efficient and simpler way to use mutation for multi objective optimization problems with mixed variables? and what is the default probability value and mutation method, if I use my code without any mutation setting (algorithm = MixedVariableGA(pop_size=110,  survival= RankAndCrowding())?


`

                mutation = {
                          Real: PM(prob=mutation_prob, eta=mutation_eta),
                          Integer: PM(vtype=float, repair=RoundingRepair(), prob=mutation_prob)
                          }
               algorithm = MixedVariableGA(pop_size=110,
                           survival= RankAndCrowding(),
                           mutation=mutation,
                           selection=selection,
                           crossover=crossover)

`

I also defined a custom mutation class and set it in MixedVariableGA, but this class does not call by optimization algorithm.

**Reply by blankjul** [reply]
I performance is an issue, I recommend implementing your own mating supporting the variables you are using.
Checkout this guide: https://pymoo.org/customization/custom.html
